File: Task 2/Java/logistics/src/main/java/com/logistics/Components/Product/Decor.java
package com.logistics.Components.Product;

import java.util.Calendar;

public class Decor implements Furniture {
    private float cost;
    private String name;
    private String supplier;
    private float discountAmount;
    private Packaging packaging;
    private String material;
    private String style;

    public Decor() {
        cost = 0;
        name = "";
        supplier = "";
        discountAmount = 1;
        packaging = new Packaging(0, 0);
        material = "";
        style = "";
    }

    public Decor(String name, float cost, String supplier, float discountAmount, int packingUnits, float costPerUnit, String material, String style) {
        this.name = name;
        this.cost = cost;
        this.supplier = supplier;
        this.discountAmount = discountAmount;
        this.packaging = new Packaging(packingUnits, costPerUnit);
        this.material = material;
        this.style = style;
    }

    // Copy Constructor
    public Decor(Decor other) {
        this.name = other.name;
        this.cost = other.cost;
        this.supplier = other.supplier;
        this.discountAmount = other.discountAmount;
        this.packaging = new Packaging(other.packaging);
        this.material = other.material;
        this.style = other.style;
    }

    // Clone method
    @Override public Decor clone() {
        return new Decor(this);
    }

    // Getters
    @Override public float getCost() { return cost; }
    @Override public String getName() { return name; }
    @Override public String getSupplier() { return supplier; }
    @Override public float getDiscountAmount(){return discountAmount;}
    @Override public float getPackagingCost() { return packaging.calculateTotalCost(); }
    @Override public String getMaterial() { return material; }
    public String getStyle() { return style; }

    // Setters
    @Override public void setCost(float cost) { this.cost = cost; }
    @Override public void setName(String name) { this.name = name; }
    @Override public void setSupplier(String supplier) { this.supplier = supplier; }
    @Override public void setDiscountAmount(float discountAmount){this.discountAmount = discountAmount;}
    @Override public void setPackagingData(int packingUnits, float costPerUnit) { this.packaging = new Packaging(packingUnits, costPerUnit); }
    @Override public void setMaterial(String material) { this.material = material; }
    public void setStyle(String style) { this.style = style; }

    @Override public float calculateDiscount() {
        // Get the current month
        int currentMonth = Calendar.getInstance().get(Calendar.MONTH);

        // Check if the current month is December 
        if (currentMonth == Calendar.DECEMBER && (discountAmount >= 0 && discountAmount <= 1)) {
            return (1 - discountAmount);
        } else {
            return 1;
        }
    }

    @Override public String display_product_info() {
        return "\nName: " + name +
               "\nCost: " + cost +
               "\nSupplier: " + supplier +
               "\nDiscount Amount: " + calculateDiscount() +
               "\nPackaging Cost: " + getPackagingCost() + 
               "\nMaterial: " + material + 
               "\nStyle: " + style;
    }
}

File: Task 4/C++Stock Viewer/Product/Decor.cpp
#include "Decor.h"

Decor::Decor() : Furniture(){
    style = "";
}

Decor::Decor(std::string name, float cost, std::string supplier, float discount_amount, int packing_units, float cost_per_unit, std::string material, std::string style)
        : Furniture(name, cost, supplier, discount_amount, packing_units, cost_per_unit, material){
    this->style = style;
}

Decor::Decor(const Decor& other) : Furniture(other) {
    this->style = other.style;
}


std::string Decor::display_product_info(){
    return Furniture::display_product_info() +
           "\nStyle: " + style;
}

// ========================================== PROTOBUFFERS
void Decor::copyToProto(PBStockItem* message){
    PBDecor* pbDecor = message->mutable_decor();
    pbDecor->set_style(style);

    PBFurniture* pbFurniture = pbDecor->mutable_furniture();
    pbFurniture->set_material(material);

    PBProduct* pbProduct = pbFurniture->mutable_product();
    pbProduct->set_name(name);
    pbProduct->set_cost(cost);
    pbProduct->set_supplier(supplier);

    PBPackage* pbPackage = pbProduct->mutable_packaging();
    pbPackage->set_units(packaging.getUnits());
    pbPackage->set_costperunit(packaging.getCost_Per_Unit());
}

void Decor::copyFromProto(const PBStockItem& message) {
    const PBDecor& pbDecor = message.decor();
    name = pbDecor.furniture().product().name();
    cost = pbDecor.furniture().product().cost();
    supplier = pbDecor.furniture().product().supplier();
    packaging.setUnits(pbDecor.furniture().product().packaging().units());
    packaging.setCost_Per_Unit(pbDecor.furniture().product().packaging().costperunit());
    material = pbDecor.furniture().material();
    style = pbDecor.style();
}